nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,529 @@
+
+        if dtype is None:
+            dtype = {x: all_dtypes[graph.index.entry_id(x)] for x in graph.index.input_names}
+
+    if None in dtype.values():
+        raise ValueError("Input variables with no type: {}".format(dtype))
+
+    compute_graph, lib, _ = nnvm.compiler.build(graph, target, shape=shape, dtype=dtype)
+    module = graph_runtime.create(compute_graph, lib, ctx)
+
+    all_shapes = compute_graph.json_attr('shape')
+    all_dtypes = [TCODE_TO_DTYPE[dtype] for dtype in compute_graph.json_attr('dtype')]
+
+    def run(**kwargs):
+        module.run(**kwargs)
+        res = []
+        for i, out_entry in enumerate(compute_graph.index.output_entries):
+            res.append(
+                module.get_output(
+                    i, tvm.nd.empty(all_shapes[out_entry[0]], all_dtypes[out_entry[0]])).asnumpy())
+        return res
+
+    return run
+
+def _dict_var_to_dict_str(dictionary):
+    """Convert a Dict[nnvm.Symbol, T] to Dict[str, T]"""
+    if isinstance(dictionary, dict):
+        return {s.attr('name') if isinstance(s, nnvm.symbol.Symbol) else s:
+                dictionary[s] for s in dictionary}
+    else:
+        return dictionary
+
+def check_function(symbol, forward=None, backward=None, grad_input_vars=None,
+                   shape=None, dtype=None, in_range=None, values=None,
+                   exclude_targets=None, only_targets=None,
+                   additional_params=None,
+                   numerical_grads=None, numerical_grads_params=None,
+                   atol=1e-5, rtol=1e-5, quiet=False):
+    """Compute the function and/or its gradients on a random input and raise
+    an exception if the result doesn't match the reference implementation.
+
+    Parameters
+    ----------
+    symbol : nnvm.Symbol
+        A symbol representing the output.
+
+    forward : Callable[..., List[numpy.ndarray]], optional
+        A reference implementation to compare with.
+
+    backward : Callable[..., List[numpy.ndarray] or Dict[str, numpy.ndarray]], optional
+        A reference implementation of gradients. Should also accept head_grads besides
+        normal inputs which is a list of gradients of some scalar wrt the outputs or just a
+        single gradient if there are multiple outputs.
+        Should return either a dict mapping input variable names to the respective
+        gradients or a list of gradients wrt variables from grad_input_vars in
+        exactly the same order (in alphabetical order by default).
+
+    grad_input_vars : List[nnvm.Symbol or str], optional
+        A list of variables with respect to which the gradients will be computed.
+        None (default) means that all input variables will be used in an alphabetical order.
+
+    shape : Dict[nnvm.Symbol or str, Tuple[int]] or Tuple[int], optional
+        A dict mapping input variable names to shapes, or just a single shape.
+        By default shapes will be inferred automatically.
```

Comment:
How can we infer input shapes? I think this argument shouldn't be optional.

Reply:
They are inferred from variables' attributes:

    x = sym.Variable("x", shape=(1, 2))

Reply:
This interface doesn't have any constraint to input symbol. So I can write some codes like:
data = sym.Variable("data")
net = sym.conv2d(data, ...)
check_function(net, shape=None, ...)

It will raise exception later, since no shape information is provided? Currently this kind of input is marked as valid.

Reply:
Oh, the error message is not nice in this case. I think making it nicer and clarifying the documentation should be enough.